### ISSUE TITLE
Cannot pickle graph

### ISSUE THREAD

**Issue by saibalmars** [networkit/networkit#495]
Hi, I got some problem when I tried to pickle an nk graph:
``` python
import netwokit as nk
import pickle
g =nk.graph.Graph(2)
g.addEdge(0,1)
pickle.dumps(g)
```

This code gives me errors as follows:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3325, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-11-79ae2e0d8bd4>", line 1, in <module>
    pickle.dumps(g)
  File "stringsource", line 2, in networkit._NetworKit.Graph.__reduce_cython__
TypeError: no default __reduce__ due to non-trivial __cinit__
```

Is there any way to fix this? 
Or is there any way to let it be readable for multiprocessing.pool?

**Comment by avdgrinten** [comment]
That is expected. You can serialize the graph to a file using the `NetworkitBinaryWriter` and deserialize it using the `NetworkitBinaryReader` (or the `readGraph` / `writeGraph`) helper functions. Other format also work but the NetworKit binary format generally has lowest performance overhead.

**Comment by michitux** [comment]
Note also that all `run()`-methods as well as some other long-running methods in NetworKit release the GIL, therefore if you mainly want to run several algorithms implemented in NetworKit in parallel, a pool of threads (like `multiprocessing.pool.ThreadPool`) should be enough to achieve decent parallelism.

**Comment by gijzelaerr** [comment]
I just ran into the same issue. isn't it possible to integrate the NetworkitBinaryReader & Writer by implementing the pickling interfaces? Methods like:

https://docs.python.org/3/library/pickle.html#object.__getstate__

Or are there low-level reasons why this might nog work?

**Comment by AndrewHannigan** [comment]
@gijzelaerr I think currently things like edge indexes don't get written to the Networkit binary format.  So a round trip from in-memory nk graph to binary format back to in-memory nk graph does not necessarily recreate an identical graph, and you would have to rebuild the edge indexes aagain.  So that complicates a pickle interface, where you'd expect the round trip to return something identical.

**Comment by avdgrinten** [comment]
It would be possible (and desirable) to extend the NetworKit binary format to also support edge indices though. The binary format stores a feature bitmask; it would be possible to add this without sacrificing backwards compatibility.

**Comment by AndrewHannigan** [comment]
Agreed and there are benefits there beyond just pickling interface.  Indexing edges can take a while, and if you can do that once and write it out as part of the binary format that saves a lot of time whenever you reload the graph since you don't have to reindex the edges.

**Comment by avdgrinten** [comment]
Alright, maybe that's actually something that @Relux-the-Relux can look into (after discussing how the file format should be augmented).

**Comment by AndrewHannigan** [comment]
Is there anything else about the state of a graph object that isn't currently captured in the binary format besides the edge index?

**Comment by opatut** [comment]
It appears this has been merged into 9.1.1, but I can't seem to get it to work. Is there still an issue with it or am I doing it wrong?

```python
import networkit, pickle

print("Networkit version:", networkit.__version__)

g = networkit.graph.Graph(directed=True, edgesIndexed=True)
g.addNode()
g.addNode()
g.addEdge(0, 1)
g.addEdge(0, 1)
print("Before serialization")
g.forEdgesOf(0, print)

networkit.graphio.writeGraph(g, "a.graph", networkit.Format.NetworkitBinary)
g2 = networkit.readGraph("a.graph", networkit.Format.NetworkitBinary)
print("Read from file")
g2.forEdgesOf(0, print)

print("Unpickled")
g3 = pickle.loads(pickle.dumps(g))
g3.forEdgesOf(0, print)
```

Output is:


```
Networkit version: 9.1.1
Before serialization
0 1 1.0 0
0 1 1.0 1
Read from file
0 1 1.0 0
0 1 1.0 0
Unpickled
0 1 1.0 0
0 1 1.0 0
```

After reading or unpickling, the last column (edge index) gives me 0 only. Should it not give me back the indexes as before serialization?

Thanks for any help.
